### PRAKTIKUM WEEK 4 SESI 3
### NAMA : Monica Putri Lestari Panjaitan
### NIM : 42324008
### PRODI : D3 TEKNOLOGI INFORMASI

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris

# Load dataset Iris dari scikit-Learn
iris = load_iris()

### Pada langkah ini, saya mengimpor library numpy dan pandas yang digunakan untuk pengolahan data numerik dan tabel selanjutnya, saya memuat sataset Iris dari pustaka scikit-learn dan menyimpannya ke dalam variabel iris.


In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]

# Definisikan X (fitur) dan y (target/Label)
X = df.drop(columns=["Species"])
y = df["Species"]
print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


### Di sini, saya mengonversi dataset mentah Iris menjadi bentuk DataFrame agar struktur datanya menjadi rapi dan mudah dibaca.

### Saya menambahkan kolom Species yang berisi nama-nama jenis bunga berdasarkan nilai targetnya.

### Kemudian, saya memisahkan dataset menjadi dua variabel utama, yaitu X untuk atribut/fitur ukuran bunga dan y untuk label/target spesies bunga.

In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

### Pada tahap ini, saya memanggil variabel y untuk menampilkan daftar label spesies yang ada pada dataset.

### Langkah ini bertujuan untuk memastikan bahwa data target telah terpisah dengan benar dari variabel X.

In [4]:
from sklearn.preprocessing import LabelEncoder

lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

### Pada kode ini, saya menggunakan LabelEncoder untuk mengubah data label spesies yang awalnya berupa teks (string) menjadi format angka numerik (0, 1, 2).

### Perubahan ini dilakukan karena model pembelajaran mesin dapat memproses data dalam bentuk angka dengan lebih efisien.

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

### Di sini, saya membagi dataset menjadi dua bagian menggunakan train_test_split, yaitu data latih (training) dan data uji (testing).

### Pembagian data dilakukan dengan proporsi 80% untuk pelatihan model (X_train, y_train) dan 20% untuk pengujian model (X_test, y_test).

In [6]:
from sklearn.naive_bayes import GaussianNB

model = GaussianNB()
NB_model = model.fit(X_train, y_train)

### Pada tahap ini, saya mengimpor kelas GaussianNB dari pustaka sklearn.naive_bayes yang cocok digunakan untuk fitur bernilai numerik kontinu.   
### Saya membuat objek model klasifikasi dengan memanggil fungsi GaussianNB() dan menyimpannya ke dalam variabel model.   
### Selanjutnya, saya melatih model tersebut menggunakan metode .fit() dengan memasukkan data latih (X_train dan y_train) agar algoritma dapat mempelajari pola hubungan antara fitur dan labelnya.   
### Hasil proses pelatihan ini disimpan ke dalam variabel NB_model untuk digunakan pada tahap prediksi selanjutnya.

In [7]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

### Pada tahap ini, saya melakukan prediksi terhadap data uji (X_test) menggunakan model NB_model yang telah dilatih.   
### Fungsi .predict() akan mengembalikan hasil tebakan kelas berupa array angka (0, 1, atau 2) berdasarkan pola yang dipelajari.   
### Hasil prediksi tersebut kemudian disimpan ke dalam variabel result dan ditampilkan untuk diproses lebih lanjut. 

In [8]:
from sklearn.metrics import accuracy_score

print('Accuracy of data testing: ', accuracy_score(y_test, result) * 100)

Accuracy of data testing:  96.66666666666667


### Selanjutnya Saya menghitung tingkat akurasi model dengan membandingkan nilai label asli (y_test) terhadap hasil prediksi (result).   
### Nilai akurasi dikalikan 100 untuk menampilkan performa model dalam bentuk persentase, yang pada pengujian ini menghasilkan akurasi sebesar 96.67%.

In [9]:
# prediksi inputan data baru
# input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\monik\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


### Pada kode ini, saya menguji kemampuan model dengan memasukkan satu sampel data baru dalam format list sederhana [[5.1, 3.5, 1.4, 0.2]].   
### Hasil prediksi berupa nilai numerik kemudian diubah kembali menjadi nama teks spesies aslinya menggunakan lab.inverse_transform().   
### Namun, karena data dikirimkan tanpa nama kolom, sistem memunculkan peringatan (warning) dari scikit-learn.  

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


### Ketika GaussianNB dilatih dengan dataset Iris, fitur-fitur dalam dataset memiliki nama kolom (iris.feature_names).   
### Namun, saat melakukan prediksi dengan NumPy array/list langsung pada langkah sebelumnya, fitur tidak memiliki nama sehingga muncul peringatan.   
### Solusinya dengan mengubah NumPy array menjadi DataFrame beserta nama kolomnya membuat struktur data konsisten dengan model yang sudah dilatih, sehingga tidak ada lagi peringatan yang muncul

### TUGAS

In [11]:
input_tugas = pd.DataFrame(np.array([[7.3, 3.8, 4.6, 2.3]]), columns=iris.feature_names)
result_tugas = NB_model.predict(input_tugas)
prediction_tugas = lab.inverse_transform(result_tugas)
print('Hasil prediksi untuk data baru adalah:', prediction_tugas)

Hasil prediksi untuk data baru adalah: ['virginica']


### Pada langkah ini, saya membuat DataFrame baru berisi data nilai [7.3, 3.8, 4.6, 2.3] beserta nama kolomnya untuk menyelesaikan tugas nomor 2.   
### Saya melakukan prediksi menggunakan model NB_model dan mengonversi hasil prediksi angkanya kembali menjadi label spesies menggunakan lab.inverse_transform().   
### Berdasarkan hasil eksekusi, sampel data baru tersebut berhasil diprediksi dan diklasifikasikan ke dalam spesies Virginica.